File: bom-analyzer/lib/cache.js
'use strict';

const fs = require('fs');
const path = require('path');

// Both supplier APIs have modest free-tier quotas (DigiKey and Mouser each cap
// a free key in the low thousands of calls per day), so every lookup that can
// be served from cache is one that stays inside the quota. Entries survive a
// restart because a BOM is usually re-uploaded several times while a user is
// fixing column mappings.
class PartCache {
  constructor(options) {
    const opts = options || {};
    this.ttlMs = opts.ttlMs || 6 * 60 * 60 * 1000;
    this.maxEntries = opts.maxEntries || 5000;
    this.file = opts.file || null;
    this.map = new Map();
    this.dirty = false;
    this.flushTimer = null;
    if (this.file) this.load();
  }

  load() {
    let raw;
    try {
      raw = fs.readFileSync(this.file, 'utf8');
    } catch (err) {
      return;
    }
    let data;
    try {
      data = JSON.parse(raw);
    } catch (err) {
      return;
    }
    if (!data || typeof data !== 'object') return;
    const now = Date.now();
    for (const [key, entry] of Object.entries(data)) {
      if (!entry || typeof entry.storedAt !== 'number') continue;
      if (now - entry.storedAt > this.ttlMs) continue;
      this.map.set(key, entry);
    }
  }

  get(key) {
    const entry = this.map.get(key);
    if (!entry) return null;
    if (Date.now() - entry.storedAt > this.ttlMs) {
      this.map.delete(key);
      return null;
    }
    // Refresh recency so the size trim evicts genuinely cold entries.
    this.map.delete(key);
    this.map.set(key, entry);
    return entry.value;
  }

  set(key, value) {
    this.map.delete(key);
    this.map.set(key, { storedAt: Date.now(), value });
    while (this.map.size > this.maxEntries) {
      const oldest = this.map.keys().next();
      if (oldest.done) break;
      this.map.delete(oldest.value);
    }
    this.scheduleFlush();
  }

  clear() {
    this.map.clear();
    this.scheduleFlush();
  }

  get size() {
    return this.map.size;
  }

  scheduleFlush() {
    if (!this.file) return;
    this.dirty = true;
    if (this.flushTimer) return;
    this.flushTimer = setTimeout(() => {
      this.flushTimer = null;
      this.flush();
    }, 2000);
    if (this.flushTimer.unref) this.flushTimer.unref();
  }

  flush() {
    if (!this.file || !this.dirty) return;
    this.dirty = false;
    const obj = {};
    for (const [key, entry] of this.map) obj[key] = entry;
    const tmp = this.file + '.tmp';
    try {
      fs.mkdirSync(path.dirname(this.file), { recursive: true });
      fs.writeFileSync(tmp, JSON.stringify(obj));
      fs.renameSync(tmp, this.file);
    } catch (err) {
      // A cache that cannot persist is still a working in-memory cache.
    }
  }
}

module.exports = { PartCache };

File: bom-analyzer/bomlib/cache.py
"""TTL cache with disk persistence.

Both supplier APIs have modest free-tier quotas, so every lookup served from
cache is one that stays inside the quota. Entries survive a restart because a
BOM is usually re-uploaded several times while a user fixes column mappings.
"""

import json
import os
import threading
import time
from collections import OrderedDict


class PartCache:
    def __init__(self, ttl_seconds=6 * 60 * 60, max_entries=5000, path=None):
        self.ttl_seconds = ttl_seconds
        self.max_entries = max_entries
        self.path = path
        self._lock = threading.Lock()
        self._entries = OrderedDict()
        if self.path:
            self._load()

    def _load(self):
        try:
            with open(self.path, 'r', encoding='utf-8') as handle:
                data = json.load(handle)
        except (OSError, ValueError):
            return
        if not isinstance(data, dict):
            return
        now = time.time()
        for key, entry in data.items():
            if not isinstance(entry, dict) or not isinstance(entry.get('storedAt'), (int, float)):
                continue
            if now - entry['storedAt'] > self.ttl_seconds:
                continue
            self._entries[key] = entry

    def get(self, key):
        with self._lock:
            entry = self._entries.get(key)
            if entry is None:
                return None
            if time.time() - entry['storedAt'] > self.ttl_seconds:
                del self._entries[key]
                return None
            # Refresh recency so the size trim evicts genuinely cold entries.
            self._entries.move_to_end(key)
            return entry['value']

    def set(self, key, value):
        with self._lock:
            self._entries[key] = {'storedAt': time.time(), 'value': value}
            self._entries.move_to_end(key)
            while len(self._entries) > self.max_entries:
                self._entries.popitem(last=False)
        self.flush()

    def clear(self):
        with self._lock:
            self._entries.clear()
        self.flush()

    def __len__(self):
        with self._lock:
            return len(self._entries)

    def flush(self):
        if not self.path:
            return
        with self._lock:
            snapshot = dict(self._entries)
        tmp = self.path + '.tmp'
        try:
            directory = os.path.dirname(self.path)
            if directory:
                os.makedirs(directory, exist_ok=True)
            with open(tmp, 'w', encoding='utf-8') as handle:
                json.dump(snapshot, handle)
            os.replace(tmp, self.path)
        except OSError:
            # A cache that cannot persist is still a working in-memory cache.
            pass
